[MRG+1] cross_val_predict handles multi-label predict_proba

stephen-hoover opened the pull request:
#### What does this implement/fix? Explain your changes.

Fixes #11058

The following fails under v0.18.1:
```python
from sklearn.datasets import make_multilabel_classification
from sklearn.model_selection import cross_val_predict
from sklearn.ensemble import RandomForestClassifier

X, y = make_multilabel_classification(n_samples=100, n_labels=3, n_classes=4, n_features=5, random_state=42)
y[:, 0] += y[:, 1]  # Put three classes in the first column
est = RandomForestClassifier(n_estimators=5, random_state=0)
cross_val_predict(est, X, y, 'predict_proba')
```

The error is
```python
ValueError: Found input variables with inconsistent numbers of samples: [100, 100, 13]
```

This PR modifies the `cross_val_predict` and `_fit_and_predict` functions so that they handle multi-label (and also multi-class multi-label) classification problems with the `predict_proba`, `predict_log_proba`, and `decision_function` methods. 

I've found two different kinds of multi-label outputs from scikit-learn estimators. The `OneVersusRestClassifier` handles multi-label tasks with binary indicator target arrays (no multi-label targets). It outputs 2D arrays from `predict_proba`, etc. methods. The `RandomForestClassifier` handles multi-class multi-label problems. It outputs a list of 2D arrays from `predict_proba`, etc.

I recognize the RandomForest-like outputs by type-checking. Lists of 2D arrays require slightly different code for keeping track of indices than single 2D output arrays.

#### Any other comments?

I didn't make any modifications to handle sparse outputs for these cases. I don't know if it's necessary. Do any estimators return sparse outputs for multi-class multi-label classification problems?

Commit:
ENH cross_val_predict and _fit_and_predict handle multi-label

Modify the `cross_val_predict` and `_fit_and_predict` functions so that they handle multi-label (and multi-class multi-label) classification problems with `predict_proba`, `predict_log_proba`, and `decision_function` methods.

There's two different kinds of multi-label outputs from scikit-learn estimators. The `OneVersusRestClassifier` handles multi-label tasks with binary indicator target arrays (no multi-label targets). It outputs 2D arrays from `predict_proba`, etc. methods. The `RandomForestClassifier` handles multi-class multi-label problems. It outputs a list of 2D arrays from `predict_proba`, etc.

Recognize the RandomForest-like outputs by type-checking. Lists of 2D arrays require slightly different code for keeping track of indices.

jnothman commented:
Sparse outputs aren't so relevant to `predict_proba`

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -393,9 +393,18 @@ def cross_val_predict(estimator, X, y=None, groups=None, cv=None, n_jobs=1,
-    if method in ['decision_function', 'predict_proba', 'predict_log_proba']:
-        le = LabelEncoder()
-        y = le.fit_transform(y)
+    do_manual_encoding = method in ['decision_function', 'predict_proba',
```
I don't like this variable name. Perhaps just `encode`, `encoded` or `is_encoded` would do

stephen-hoover replied:
Changed to `encode`.

Commit:
do_manual_encoding -> encode; better commenting

Commit:
Merge remote-tracking branch 'upstream/master' into multilabel-cross-val-predict

stephen-hoover commented on sklearn/model_selection/_validation.py:
```diff
@@ -419,9 +428,20 @@ def cross_val_predict(estimator, X, y=None, groups=None, cv=None, n_jobs=1,
     # Check for sparse predictions
```
Removed.

stephen-hoover commented on sklearn/model_selection/_validation.py:
```diff
@@ -419,9 +428,20 @@ def cross_val_predict(estimator, X, y=None, groups=None, cv=None, n_jobs=1,
     if sp.issparse(predictions[0]):
         predictions = sp.vstack(predictions, format=predictions[0].format)
+    elif do_manual_encoding and isinstance(predictions[0], list):
+        n_labels = y.shape[1]
```
Added a comment.

stephen-hoover commented on sklearn/model_selection/_validation.py:
```diff
@@ -419,9 +428,20 @@ def cross_val_predict(estimator, X, y=None, groups=None, cv=None, n_jobs=1,
         predictions = np.concatenate(predictions)
-    return predictions[inv_test_indices]
+
+    if do_manual_encoding and isinstance(predictions, list):
```
Ah, you're right. Checking if `predictions` is a list is enough. Changed.

stephen-hoover commented:
@jnothman , apologies for taking a long time to respond here. I believe I've addressed your comments.

stephen-hoover commented:
@jnothman , would you like additional changes here?

jnothman commented:
Oh, I thought we'd merged this fix at some point :(

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -416,12 +427,26 @@ def cross_val_predict(estimator, X, y=None, groups=None, cv=None, n_jobs=1,
+    elif encode and isinstance(predictions[0], list):
+        # `predictions` is a list of method outputs from each fold.
+        # If each of those is also a list, then treat this as a
+        # multi-class multi-label task. We need to separately concatenate
```
do you mean multi-output multiclass?

jnothman reviewed (commented):
I think this is good, but I'm too tired to be sure!

Commit:
Change docs to multioutput-multiclass, backticks

stephen-hoover commented:
@jnothman , thanks for taking another look at this PR. I added double backticks and changed "multi-class multi-label" to "multioutput-multiclass" (the spelling I found in existing documentation) in that comment and in the What's New.

jnothman commented:
definitely wasn't a full review yet. All of the class list games are so
ugly!

On 28 Jul 2017 2:43 am, "Stephen Hoover" <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> , thanks for taking another look
> at this PR. I added double backticks and changed "multi-class multi-label"
> to "multioutput-multiclass" (the spelling I found in existing
> documentation) in that comment and in the What's New.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8773#issuecomment-318418912>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6xTsIRXhSnVb8v6xEUNzcjWRJIcXks5sSL5EgaJpZM4NEX4n>
> .
>

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -915,54 +916,101 @@ def test_cross_val_predict_sparse_prediction():
-            expected_predictions[test] = func(X[test])
+            preds = func(X[test])
+            if isinstance(predictions, list):
+                for i_label in range(y.shape[1]):
```
Call it output rather than label, please

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -915,54 +916,101 @@ def test_cross_val_predict_sparse_prediction():
         for train, test in kfold.split(X, y):
             est.fit(X[train], y[train])
-            expected_predictions[test] = func(X[test])
+            preds = func(X[test])
```
I'm a bit confused. Doesn't this mean that sometimes we're going to have mismatched numbers of classes vis-a-vis what we try to solve with `_enforce_prediction_order`?

jnothman reviewed (commented):
I'm not convinced that the need for `_enforce_prediction_order` is tested here. Can we make sure this is tested on data where the set of classes for training each fold will vary, such as `y = [1,2,3,4,5]` and `Y = [[0, 1], [1, 2], [0, 3], [1, 4], [0, 5]]`?

It wouldn't hurt if `check_cross_val_predict_with_method` explicitly checked that the shape was as expected, either...

Commit:
TST Move for loop out of check_cross_val_predict_with_method

Let individual test functions control the for loops.

Commit:
TST Test shape of cross_val_predict outputs

Commit:
TST Rename i_label -> i_output

Commit:
TST Test cases where _enforce_prediction_order is needed

stephen-hoover commented:
@jnothman , you're right that the tests weren't covering the case where `_enforce_prediction_order` was needed. I added two new tests which have a split with fewer classes than the full dataset. I also added some explicit asserts on the output.

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -480,16 +505,42 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+    then the output prediction array might not have the same
+    columns as other folds. Use the list of class names
+    (assumed to be integers) to enforce the correct column order.
+    """
```
Is it worth fast-pathing the classes == arange case? Or is that premature optimisation?

stephen-hoover replied:
That's a good idea. We can return immediately if all classes were present in the subset of data used to train this fold.

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -480,16 +505,42 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+    predictions_ = np.zeros((predictions.shape[0], n_classes),
+                            dtype=predictions.dtype)
+    if one_col_if_binary and len(classes) == 2:
+        predictions_[:, classes[-1]] = predictions
```
This leaves one of the columns zeroed?

stephen-hoover replied:
That's what we have to do if `n_classes >= 3`. But you're right that this is a problem if `n_classes = 2`. That should have a single column of output. I added a new test to catch this, `test_cross_val_predict_binary_decision_function`. It's handled now by having the function return `predictions` directly when `len(classes) == n_classes`.

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -915,54 +916,157 @@ def test_cross_val_predict_sparse_prediction():
+
+
+def test_cross_val_predict_with_method_multilabel_rf():
+    # The RandomForest allows anything for the contents of the labels.
```
The wording here is unclear. Do you just mean that RF handles multiclass-multioutput and produces `predict_proba` in that vein?

stephen-hoover replied:
Changed to "The RandomForest allows multiple classes in each label.".

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -915,54 +916,157 @@ def test_cross_val_predict_sparse_prediction():
+    X = rng.normal(0, 1, size=(10, 10))
+    y = np.array([0, 1, 0, 1, 0, 1, 0, 1, 0, 2])
+    est = LogisticRegression()
+    for method in ['predict_proba', 'predict_log_proba']:
```
decision_function also?

stephen-hoover replied:
I'd left it off because the test code doesn't handle the decision_function case when there's three classes in the full data but only 2 in one of the folds. I think that case is covered by `test_cross_val_predict_class_subset`. In this test, I modified `y` to have 4 classes instead of 3 and added "decision_function" to the list.

Commit:
BUG Output one column for binary decision_function

Also don't do unnecessary work if number of classes in a cross_val_predict fold equals number of classes in the full data.

stephen-hoover commented:
@jnothman , comments addressed. Thanks for suggesting that optimization and pointing out a bug in `_enforce_prediction_order`.

amueller commented:
needs merge fix

Commit:
Merge remote-tracking branch 'upstream/master' into multilabel-cross-val-predict

Commit:
flake8

stephen-hoover commented:
Thank you for flagging, @amueller ! I merged the master branch in.

amueller commented on doc/whats_new.rst:
```diff
@@ -416,6 +416,10 @@ Miscellaneous
 - :class:`dummy.DummyClassifier` and :class:`dummy.DummyRegressor`
   now accept non-finite features. :issue:`8931` by :user:`Attractadore`.
 
+   - Added ability for :func:`model_selection.cross_val_predict` to handle multi-label
```
indentation error

jnothman replied:
Given its relation to other fixes we'll put in [0.19.1](https://github.com/scikit-learn/scikit-learn/milestones/0.19.1), I think we should include this in the bug-fix release.

Commit:
Merge remote-tracking branch 'upstream/master' into multilabel-cross-val-predict

Commit:
TST Adjust tests from merge

Commit:
DOC Fix What's New entry

amueller commented:
gonna review now, but I'm really unhappy that we have to special case for multioutput-multiclass everywhere. In particular having the user-interface change between estimators seems very un-sklearn like. Opened #9650 to vent.

amueller commented:
@jnothman what's the motivation for putting this into 0.19.1? This is not a regression, right?

amueller commented on sklearn/model_selection/_validation.py:
```diff
@@ -639,9 +639,20 @@ def cross_val_predict(estimator, X, y=None, groups=None, cv=None, n_jobs=1,
+    # we need to manually encode classes to ensure consistent column ordering.
+    encode = method in ['decision_function', 'predict_proba',
+                        'predict_log_proba']
+    if encode:
```
These changes conflict with #9593, right?

stephen-hoover replied:
If I read #9593 correctly, the change in https://github.com/scikit-learn/scikit-learn/pull/8773/files#diff-60033c11a662f460e1567effd5faa6f0R784 (the first line of `_enforce_prediction_order`) is the same as the code change in that PR.

amueller commented on sklearn/model_selection/_validation.py:
```diff
@@ -639,9 +639,20 @@ def cross_val_predict(estimator, X, y=None, groups=None, cv=None, n_jobs=1,
+            y = le.fit_transform(y)
+        elif y.ndim == 2:
+            y_enc = np.zeros_like(y, dtype=np.int)
+            for i_label in range(y.shape[1]):
```
This seems redundant in the multi-label case (not multioutput multiclass)

stephen-hoover replied:
Yes, it is. The same redundancy is present in the original 1D case, for what it's worth. The `y` gets encoded and transformed whether it needs it or not. It does seem more costly in the 2D case to run the encoding when it's not needed. 

The only way I'm thinking to check whether we're in a multi-label case is to run a `set(y.flat) == {0, 1}` check, or maybe `np.unique(y).tolist() == [0, 1]`. Do you think the performance tradeoff is worth it?

amueller reviewed (commented):
I'm not sure this is the right approach. I feel a more general solution like discussed in #8100 would be good.

jnothman commented:
I don't mind whether or not it's in 0.19.1. but it may be appropriate
alongside a series of fixes to cross_val_predict. I believe that if we
don't put this in 0.19.1 we should ideally improve the error messages in
this case for 0.19.1; so why not just fix it?

On 31 Aug 2017 1:21 am, "Andreas Mueller" <notifications@github.com> wrote:

> *@amueller* commented on this pull request.
>
> I'm not sure this is the right approach. I feel a more general solution
> like discussed in #8100
> <https://github.com/scikit-learn/scikit-learn/issues/8100> would be good.
> ------------------------------
>
> In sklearn/model_selection/_validation.py
> <https://github.com/scikit-learn/scikit-learn/pull/8773#discussion_r136101009>
> :
>
> > @@ -639,9 +639,20 @@ def cross_val_predict(estimator, X, y=None, groups=None, cv=None, n_jobs=1,
>
>      cv = check_cv(cv, y, classifier=is_classifier(estimator))
>
> -    if method in ['decision_function', 'predict_proba', 'predict_log_proba']:
> -        le = LabelEncoder()
> -        y = le.fit_transform(y)
> +    # If classification methods produce multiple columns of output,
> +    # we need to manually encode classes to ensure consistent column ordering.
> +    encode = method in ['decision_function', 'predict_proba',
> +                        'predict_log_proba']
> +    if encode:
>
> These changes conflict with #9593
> <https://github.com/scikit-learn/scikit-learn/pull/9593>, right?
> ------------------------------
>
> In sklearn/model_selection/_validation.py
> <https://github.com/scikit-learn/scikit-learn/pull/8773#discussion_r136101238>
> :
>
> > @@ -639,9 +639,20 @@ def cross_val_predict(estimator, X, y=None, groups=None, cv=None, n_jobs=1,
>
>      cv = check_cv(cv, y, classifier=is_classifier(estimator))
>
> -    if method in ['decision_function', 'predict_proba', 'predict_log_proba']:
> -        le = LabelEncoder()
> -        y = le.fit_transform(y)
> +    # If classification methods produce multiple columns of output,
> +    # we need to manually encode classes to ensure consistent column ordering.
> +    encode = method in ['decision_function', 'predict_proba',
> +                        'predict_log_proba']
> +    if encode:
> +        y = np.asarray(y)
> +        if y.ndim == 1:
> +            le = LabelEncoder()
> +            y = le.fit_transform(y)
> +        elif y.ndim == 2:
> +            y_enc = np.zeros_like(y, dtype=np.int)
> +            for i_label in range(y.shape[1]):
>
> This seems redundant in the multi-label case (not multioutput multiclass)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8773#pullrequestreview-59592873>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz61G432NFoCTMQodZYZT_pzAL5GqEks5sdX4NgaJpZM4NEX4n>
> .
>

stephen-hoover commented:
I think that all the special-casing in this PR is really ugly, but it fixes the immediate problem of `cross_val_predict` not working on 2D target arrays. Global fixes like the ones discussed in #8100 and #9650 sound like better long-term solutions, but I expect they'll also take a lot longer to complete.

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1151,61 +1152,194 @@ def test_cross_val_predict_sparse_prediction():
+                if preds[i_out].ndim == 1 and \
+                        expected_predictions[i_out].ndim == 1:
+                    # Get here for binary decision functions
+                    expected_predictions[i_out][test] = preds
```
This doesn't have test coverage?

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1151,61 +1152,194 @@ def test_cross_val_predict_sparse_prediction():
+    for train, test in kfold.split(X, y):
+        est.fit(X[train], y[train])
         func = getattr(est, method)
+        preds = func(X[test])
```
I don't like the simultaneous presence of preds and predictions...

jnothman reviewed (commented):
I think we could consider merging this instead of #9593. But we'd need to check that all the things handled there are here too, notably `SVC(decision_function_shape='ovo')`.

I'm also not very happy with `check_cross_val_predict_with_method`. The reference implementation for testing needs to be simpler. One option may be to make it use an unrolled KFold(2) and hence avoid the loop.

I don't really like the fact that it has similar handling of mismatched classes. One option may be to use sample_weight to spoof the disappearing class issues (i.e. spoof an additional sample in each class for each fit which has `sample_weight=0`), thus creating all the necessary columns but not affecting the model.

Basically the test is unreadable. We need to change that.

amueller commented:
you need to resolve the whats_new conflict by moving the changes to the correct file (the whatsnew for 0.19.1)

stephen-hoover commented:
What file is that? I'm not seeing anything obvious in the `doc/` directory. I checked other PRs on the 0.19.1 milestone, and I didn't see any merged PRs adding any changelog notes. I could remove the What's New entry entirely.

I know I still have to-dos here. I'll fix up the tests this week.

Commit:
TST Refactor check_cross_val_predict_with_method test helper

Commit:
BUG cross_val_predict handle 1-class decision function

Bugfix from #9593; handle cases when there's two classes and we're using `decision_function`, but our CV fold only sees one of the classes. Doesn't include the `ValueError` added in #9593.

Commit:
TST Test multioutput-multiclass decision functions

Commit:
DOC Remove whatsnew entry

stephen-hoover commented:
@jnothman , @amueller , I've refactored the `check_cross_val_predict_with_method` helper; hopefully it's more readable now. There's no multioutput-multilabel models in scikit-learn now which use a `decision_function`, so I added a mock for testing. If you knew there would never be a `decision_function` which provided list output, then the tests could be a bit simpler.

I copied over the tests from #9593 (except for the `ValueError` check) to see if the bug addressed there were still present here. It was, so I added a fix. I'd be happy to rebase on that PR after it merges.

It looks like #9593 might be changing `cross_val_predict` to error if not all training set classes are present in one of the CV folds. If that merges, that would allow the code changes and tests I'm making here to be simpler. Having to deal with the possibility of different training classes in each fold adds complexity to `cross_val_predict`.

Finally, I removed the "What's New" entry, since I don't see a spot where 0.19.1 changes are being recorded.

jnothman commented:
Personally, I think having classes missing from training is something that happens not so infrequently in practice. What would you prefer as a user considering the use cases of:
* visualising prediction distributions
* summarising prediction distributions
* predictions as the output of a layer of a stacked ensemble.

Would you find an exception or blank output (perhaps with a warning) more useful?

stephen-hoover commented:
Personally, I think that a blank output with a warning would be better. However, I do think that missing classes in a validation fold's training set indicates a problem. It could happen if a user input a sorted training set and didn't use a shuffled CV, which would cause the outputs to be not useful. I've done that by accident before.

reiinakano commented:
> It looks like #9593 might be changing cross_val_predict to error if not all training set classes are present in one of the CV folds. If that merges, that would allow the code changes and tests I'm making here to be simpler. Having to deal with the possibility of different training classes in each fold adds complexity to cross_val_predict.

Hello, #9593 only errors out when the output of the prediction method does not match the number of classes used to train the estimator. As far as I know, this only happens with an OvO `SVC`. It still tries to handle the special case of missing training set classes for other more "regular" estimators. 

But I agree, just erroring out when a validation fold is missing classes would make `cross_val_predict` **much** simpler and would give a useful message to beginners that imbalanced validation folds is a problem, instead of waiting for them to figure out why their probabilities don't make much sense.

jnothman commented:
well I'd think a zero probability for a rare class does make some sense
(with a warning). Similar output for decision_function might be harder to
justify.

amueller commented:
can you resolve conflicts please? Sorry for the slow turn-around

stephen-hoover commented:
Sure thing, I'll take a look at conflicts this week.

Commit:
Merge remote-tracking branch 'upstream/master' into multilabel-cross-val-predict

Commit:
BUG Preserve dtype when handling missing classes

np.full defaults to using `dtype=np.array(fill_value).dtype`, which for "predict_proba" (fill_value=0) is an int. Make the empty array here have the same data type as the predictions it's meant to hold.

Commit:
DOC Add What's New entry

jnothman commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-862dac28db5fd6418738c6fce9bbb8c55ef91855)

**new alerts:**

* 1 for Unused argument in a formatting call

---

*Comment posted by [lgtm.com](https://lgtm.com)*

Commit:
BUG Remove unused formatting argument

stephen-hoover commented on sklearn/model_selection/_validation.py:
```diff
@@ -754,53 +779,75 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+        predictions_for_all_classes = np.full((_num_samples(predictions),
+                                               n_classes),
+                                              default_values[method],
+                                              dtype=predictions.dtype)
```
This new `dtype` argument fixes a bug I found when a couple of new tests failed after merging in master. `np.full` infers the data type from the fill value, and when the method is "predict_proba", the fill value is an `int`. This caused probabilities to be set to zero for all columns of all rows in folds where one of the classes was missing from the training set.

jnothman replied:
oh!

stephen-hoover commented:
I merged in the changes from #9593 and added a What's New entry for v0.20. Tests are passing and I think that this PR is ready to go again!

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -754,53 +779,75 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+
+        float_min = np.finfo(predictions.dtype).min
+        default_values = {'decision_function': float_min,
+                          'predict_log_proba': float_min,
```
I wonder if we need to make this -np.inf for correspondence with `predict_proba`.

stephen-hoover replied:
The idea of returning infinities from something that's not exactly an error makes me nervous. I think that numpy does sensible things with an `np.inf`, so it would probably be fine. On the other hand,
```python
In [1]: 10 ** (np.finfo(float).min)
Out[1]: 0.0
```
so numerically the float minimum and negative infinity seem similar already.

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -754,53 +779,75 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+                                    predictions.shape, method, len(classes)))
+            if len(classes) <= 2:
+                # In this special case, `predictions` contains a 1D array.
+                raise ValueError('Only {} class/es in training fold, this '
```
"Only {} class/es in training but {} in overall dataset"

Should we be filling in `np.transpose([-decision_function, decision_function])`??

stephen-hoover replied:
Changed. Wouldn't we get to this branch if the full training data had, say, 3 or 4 classes, and we ended up with 1 or 2 in a fold? We'd end up with at least two blank columns to fill. Erroring seems safer than trying to make something methodologically sensible for this case.

jnothman replied:
Okay

jnothman reviewed (commented):
I need still to take a fresh look at tests.

We've had lots of trouble with `decision_function` and 
TBH it doesn't really make a lot of sense to be using `cross_val_predict` on it, because it does not normalize in a way that makes comparisons across different classifiers/training sets meaningful. Users should probably be warned not to do this...

Commit:
ENH Improve error message

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1259,61 +1261,220 @@ def test_cross_val_predict_sparse_prediction():
+    X, y = shuffle(X, y, random_state=0)
+    for method in ['decision_function', 'predict_proba', 'predict_log_proba']:
+        est = LogisticRegression()
+        check_cross_val_predict_with_method_multiclass(est, X, y, method)
```
perhaps also try this with string `y`?

stephen-hoover replied:
The `check_cross_val_predict_multiclass` called here goes through multiple transformations of `y`, including casting it as a string before running `cross_val_predict`. I think the string case is covered.

jnothman reviewed (commented):
The tests are looking pretty good. I wish this could be simpler.

Commit:
TST Shorten names and remove unneeded helper

stephen-hoover commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1259,61 +1261,220 @@ def test_cross_val_predict_sparse_prediction():
-    for method in methods:
-        predictions = cross_val_predict(est, X, y, method=method)
-        assert_equal(len(predictions), len(y))
+def check_cross_val_predict_with_method_binary(est, X, y, method):
```
Done.

stephen-hoover commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1259,61 +1261,220 @@ def test_cross_val_predict_sparse_prediction():
+                           expected_predictions)
+
+
+def check_cross_val_predict_with_method_multiclass(est, X, y, method):
```
Done.

stephen-hoover commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1259,61 +1261,220 @@ def test_cross_val_predict_sparse_prediction():
+                           expected_predictions)
+
+
+def check_cross_val_predict_with_method_multilabel(est, X, y, method):
```
Done.

stephen-hoover commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1259,61 +1261,220 @@ def test_cross_val_predict_sparse_prediction():
+        for i in range(len(x)):
+            assert_array_equal(x[i], y[i])
+    else:
+        assert_array_equal(x, y)
```
This helper function was only called once. I removed it entirely and moved the only branch which was used into the function which had previously called this helper.

stephen-hoover commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1259,61 +1261,220 @@ def test_cross_val_predict_sparse_prediction():
+        probs = self.predict_proba(X)
+        if isinstance(probs, list):
+            probs = [p[:, -1] if p.shape[1] == 2 else p for p in probs]
+        elif probs.shape[1] == 2:
```
I was a bit nervous about removing this branch. This class only exists to help me test `decision_function` for list outputs, but I thought it was better to have the mock `decision_function` not be broken for non-list outputs. Would you still prefer either that this branch turn into a `ValueError`, or that I assume it will always run with list outputs?

jnothman replied:
Just put assert isinstance before and get rid of this case

stephen-hoover replied:
Okay, done.

Commit:
TST Restrict RF helper to list outputs

stephen-hoover commented:
@jnothman , I've addressed this round of comments.

jnothman commented:
I hope it was the last round from me, but I need at least a train ride to
double check.

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -754,53 +779,76 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+                n_classes=len(set(y[:, i_label])), method=method)
+                for i_label in range(len(predictions))]
+        else:
+            # A 2D y array should be a binary label indicator matrix
```
it can't be otherwise multioutput here?

stephen-hoover replied:
I'm pretty sure it can't, no. If the predictions are a single array and not a list, and there's a 2D array of targets, then the targets are binary. Multiclass multi-target problems produce lists of arrays as output.

jnothman replied:
Right, in this context. Good.

jnothman reviewed (approved):
This LGTM. Or, at least, I'm ready to hand it off to a second reviewer to check the tests are as maintainable as they're going to get!

jnothman commented:
Could this please get a second review? I know it's a hairy part of the codebase, but it'll be necessary for stacking ensembles, if nothing else.

amueller committed:
Merge branch 'master' into multilabel-cross-val-predict

jnothman commented:
I thought of this PR yesterday. I think it's essential before we merge
stacking using cross_val_predict (though there are unresolved conflicts
there). @glemaitre? @qinhanmin2014?

amueller committed:
Merge branch 'master' into multilabel-cross-val-predict

amueller commented:
does someone wanna review? otherwise I say untag.

amueller committed:
actually commit fix

adrinjalali committed:
merge upstream/master

adrinjalali committed:
merge upstream/master, again

adrinjalali reviewed (approved):
LGTM, should we merge this while the other issues/approaches move forward?

adrinjalali committed:
fix lbfgs future warning

jnothman commented:
I think merging this will be especially valuable if we review and merge StackingClassifier.

jnothman committed:
Merge branch 'master' into multilabel-cross-val-predict

jnothman commented:
Thanks @stephen-hoover, and sorry for the very slow ride!!

stephen-hoover commented:
Thank you for the merge!
